labrador/src/poly.rs

```diff
@@ -0,0 +1,332 @@
+            }
+        }
+
+        // Reduce modulo X^d + 1
+        if result_coefficients.len() > Self::DEGREE_BOUND {
+            let q_minus_1 = Zq::MAX;
+            let (left, right) = result_coefficients.split_at_mut(Self::DEGREE_BOUND);
+            for (i, &overflow) in right.iter().enumerate() {
+                left[i] += overflow * q_minus_1;
+            }
+            result_coefficients.truncate(Self::DEGREE_BOUND);
+        }
+        PolyRing::new(result_coefficients)
+    }
+
+    /// Generate random Zq vector with a provided cryptographically secure RNG
+    pub fn random<R: Rng + CryptoRng>(rng: &mut R, n: usize) -> Self {
+        let uniform = Uniform::new_inclusive(Zq::ZERO, Zq::MAX).unwrap();
+        let mut coeffs = Vec::with_capacity(n);
+        coeffs.extend((0..n).map(|_| uniform.sample(rng)));
+        Self { coeffs }
+    }
+
+    pub fn to_rq<const D: usize>(&self) -> Rq<D> {
+        Rq::from_vec(self.get_coeffs().clone())
+    }
+
+    /// Compute the conjugate automorphism \sigma_{-1} of vector based on B) Constraints..., Page 21.
+    pub fn conjugate_automorphism(&self) -> PolyRing {
+        let q_minus_1 = Zq::MAX;
+        let mut new_coeffs = vec![Zq::ZERO; PolyRing::DEGREE_BOUND];
+        for (i, new_coeff) in new_coeffs
+            .iter_mut()
+            .enumerate()
+            .take(PolyRing::DEGREE_BOUND)
+        {
+            if i < self.get_coeffs().len() {
+                if i == 0 {
+                    *new_coeff = self.get_coeffs()[i];
+                } else {
+                    *new_coeff = self.get_coeffs()[i] * q_minus_1;
+                }
+            } else {
+                *new_coeff = Zq::ZERO;
+            }
+        }
+        let reversed_coefficients = new_coeffs
+            .iter()
+            .take(1)
+            .cloned()
+            .chain(new_coeffs.iter().skip(1).rev().cloned())
+            .collect::<Vec<Zq>>();
+
+        PolyRing::new(reversed_coefficients)
+    }
+
+    /// Compute the operator norm of a polynomial given its coefficients.
+    /// The operator norm is defined as the maximum magnitude of the DFT (eigenvalues)
+    /// of the coefficient vector.
+    pub fn operator_norm(coeffs: &[f64]) -> f64 {
+        let n = coeffs.len();
+        let mut planner = FftPlanner::new();
+        let fft = planner.plan_fft_forward(n);
+
+        // Convert coefficients into complex numbers (with zero imaginary parts)
+        let mut buffer: Vec<Complex<f64>> =
+            coeffs.iter().map(|&x| Complex { re: x, im: 0.0 }).collect();
+
+        // Compute the FFT (this gives the eigenvalues of the circulant matrix)
+        fft.process(&mut buffer);
+
+        // Return the maximum absolute value (norm) among the eigenvalues
+        buffer
+            .iter()
+            .map(|c| c.norm())
+            .fold(0.0, |max, x| max.max(x))
+    }
+
+    pub fn zq_to_f64(zq: &PolyRing) -> Vec<f64> {
```

Comment:
Probably, as `zq_to_f64 ` goes together with `operator_norm`, it's worth to put its implementation inside the `operator_norm` (thus reducing unnecessary allocations and iterations) altogether with `Zq.to_f64` to prevent leaking away floating point representations. The only thing we need from floats is the norm of the polynomial.